analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs

```diff
@@ -183,11 +183,20 @@ private ProgramState ProcessDebugAssertBoolSymbol(ProgramState state, IOperation
         {
             ObjectConstraint constraint when constraint == ObjectConstraint.Null => new[] { context.SetOperationConstraint(BoolConstraint.True) },
             ObjectConstraint constraint when constraint == ObjectConstraint.NotNull => new[] { context.SetOperationConstraint(BoolConstraint.False) },
+            _ when ConversionOperandType(invocation.Arguments[0].ToArgument().Value) is { } type && !CanBeNull(type) => new[] { context.SetOperationConstraint(BoolConstraint.False) },
             _ when invocation.Arguments[0].TrackedSymbol() is { } argumentSymbol => new[]
             {
                         context.SetOperationConstraint(BoolConstraint.True).SetSymbolConstraint(argumentSymbol, ObjectConstraint.Null),
                         context.SetOperationConstraint(BoolConstraint.False).SetSymbolConstraint(argumentSymbol, ObjectConstraint.NotNull),
             },
             _ => new[] { context.State }
         };
+
+    private static ITypeSymbol ConversionOperandType(IOperation operation) =>
+        operation.Kind == OperationKindEx.Conversion
+            ? ConversionOperandType(IConversionOperationWrapper.FromOperation(operation).Operand)
+            : operation.Type;
+
+    private static bool CanBeNull(ITypeSymbol type) =>
+        type.IsReferenceType || type.OriginalDefinition.Is(KnownType.System_Nullable_T);
```

Comment:
I added something related in #6129 (in my opinion `Nullable<T>` and unconstraint generic types should both be handled the same way here (`CanBeNull(ITypeSymbol type) => type.IsReferenceType || type.OriginalDefinition.Is(KnownType.System_Nullable_T || type.IsUnconstraintGeneric()`)

https://github.com/SonarSource/sonar-dotnet/pull/6129/files#diff-cd5d22e03a71e83850d57d225c1b12567c750d817f710f15d01abc2746a72692R43-R45

 Please add this to analyzers/src/SonarAnalyzer.Common/Helpers/TypeHelper.cs so we have a centralized place for it.

Reply:
`type.OriginalDefinition.Is(KnownType.System_Nullable_T)` There is also `IsNullableValueType` and `IsNonNullableValueType` scattered around in the codebase. These should also be centralized.

Reply:
I don't see a test for the nullable part. Am I missing something?

Reply:
For the unconstraint generics, this is not allowed. And we don't really know. I wouldn't do it unless we have a clear problem with it.
```
public static void Go<T>(T arg)
{
    arg = null;
}
```

Reply:
`IsNonNullableValueType` is in the old engine. Will be moved in case we need it, otherwise, will be deleted eventually.

Reply:
There's line 542 
```
[DataRow("Integer?")]
```
that proves that `Integer?` behaves the same as `Object` or `Exception`